woocommerce/woocommerce-gateway-stripe
# Use a non-wildcard character for Apple Pay postcodes.

This PR changes the backfill character used for Apple Pay postcodes.

Fixes #2645

## Changes proposed in this Pull Request:

As per issue 2645, Apple Pay returns a partial address for validation prior to the user confirming the transaction, and this is backfilled with `*` characters which breaks some specific workflows that rely on wildcard shipping zones.

This changes the backfill character to a different one that can be used for differentiation by the backend - e.g. separating BN1 from BN15 by way of `BN1#*` vs `BN15*` and not erroneously producing `BN1*` as a false match.

I've tested this in my staging environment and it seems to work the way I'd expect - `BN1####` seems to be validated the same general ways as `BN1****` via WC_Validation (i.e. neither is officially tracked as a GB postcode, but that's not a blocker for the process in question), but this way I can add a rule for `BN1#*` as a shipping zone. I've also tested it with other postcodes and it still seems to work normally for those.

Obviously I'd prefer to have this in the official release so I don't have to manually patch it.

## Testing instructions

* Set up two shipping zones in United Kingdom, such that `BN1#*` has one price, and `BN15*` has a different price.
* Attempt to buy things with Apple Pay with shipping addresses in BN1 and BN15 and verify that the correct shipping price is given. (In case it helps, Brighton railway station is BN1 3XP, Lancing railway station is BN15 8AG)
* Additionally, if BN1 and BN15 are given as shipping zones, verify that BN2 is denied as it is not defined as a shipping route.

There don't appear to be any unit tests currently for GB postcodes in general, and I'm not sure how far in scope it is for me to sit and write those just to get an otherwise small bug fixed.

---


**Post merge**
